File: leetcode-1_two-sum.go
/**
 * 给定一个无序int数组和一个target, 从数组中取两个元素使得和等于target
 */
package main

import "fmt"

func main() {
    fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
}

func twoSum(nums []int, target int) []int {
    m := map[int]int{}
    for k, v := range nums {
        if _, ok := m[v]; ok {
            return []int{m[v], k}
        }
        //map存储target与当前元素的差值, 当迭代元素等于map中的差值, 说明之前有元素加上当前元素等于target
        m[target - v] = k
    }
    return []int{}
}

File: leetcode-1_two-sum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * 给定一个无序 int 数组和一个 target, 从数组中取两个元素使得和等于 target，返回两个元素的下标
 */
class Solution {
    public static void main(String[] args) {
        int[] nums = {2, 7, 11, 15};
        System.out.println(Arrays.toString(twoSum(nums, 9)));;
    }

    public static int[] twoSum(int[] nums, int target) {
        // 使用 map 存储 target 和数组元素的差值和下标(差值为键，下标为值)
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            // 如果当前元素刚好是 map 中的键，说明之前有一个元素与当前元素相加刚好等于 target
            if (map.containsKey(nums[i])) {
                return new int[]{map.get(nums[i]), i};
            }
            map.put(target - nums[i], i);
        }
        return new int[0];
    }
}
